arangod/Cluster/Utils/ShardID.h

```diff
@@ -0,0 +1,161 @@
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "Basics/Exceptions.h"
+#include "Basics/ResultT.h"
+#include "Basics/StringUtils.h"
+#include "Inspection/Status.h"
+
+#include <compare>
+#include <cstdint>
+#include <fmt/format.h>
+#include <velocypack/Value.h>
+
+namespace arangodb {
+
+struct ShardID {
+
+  static ResultT<ShardID> shardIdFromString(std::string const& s) {
+    if (s.empty() || s.at(0) != 's') {
+      return Result{TRI_ERROR_BAD_PARAMETER, "Expected ShardID to start with 's'"};
+    }
+    auto res = basics::StringUtils::try_uint64(s.c_str() + 1, s.length() -1);
+    if (res.fail()) {
+      return std::move(res.result());
+    }
+    return ShardID{res.get()};
+  }
+
+  static ShardID invalidShard() { return ShardID{0}; }
+
+  // Default constructor required for Inspectors.
+  // Note: This shardID is considered Invalid.
+  ShardID() : id(0) {}
+
+  explicit ShardID(uint64_t id) : id(id) {}
+
+  explicit ShardID(std::string_view id) {
+    auto maybeShardID = shardIdFromString(std::string(id));
+    if (maybeShardID.fail()) {
+      THROW_ARANGO_EXCEPTION(maybeShardID.result());
+    }
+    *this = maybeShardID.get();
+  }
+
+
+  ~ShardID() = default;
+  ShardID(ShardID const&) = default;
+  ShardID(ShardID&&) = default;
+  ShardID& operator=(ShardID const&) = default;
+  ShardID& operator=(ShardID&&) = default;
+
+  operator std::string() const {
+    return "s" + std::to_string(id);
+  }
+
+  operator arangodb::velocypack::Value() const {
+    return arangodb::velocypack::Value("s" + std::to_string(id));
+  }
+
+  friend auto operator<=>(ShardID const&, ShardID const&) = default;
+  friend bool operator==(ShardID const&, ShardID const&) = default;
+
+  bool operator==(std::string_view other) const {
+    return other == std::string{*this};
+  }
+
+  bool operator==(std::string const& other) const {
+    return other == std::string{*this};
+  }
+
+  bool isValid() const noexcept {
+    // We can never have ShardID 0. So we use it as invalid value.
+    return id != 0;
+  }
+
+  // Add an inspector implementation, shardIDs will be serialized and deserialized as "s" + number
+  // for compatibility reasons.
+  template<class Inspector>
+  inline friend auto inspect(Inspector& f, ShardID& x) -> arangodb::inspection::Status {
+    if constexpr (Inspector::isLoading) {
+      std::string v;
+      auto res = f.apply(v);
+      if (res.ok()) {
+        auto result = shardIdFromString(v);
+        if (result.fail()) {
+          // NOTE: For some reason we cannot inline the errorMessage() into the return
+          // in that case the string_view variant will be selected, which is no valid input
+          // for status.
+          std::string msg{std::move(result.errorMessage())};
```

Comment:
The extra std::move looks redundant.
```suggestion
          std::string msg{result.errorMessage()};
```